helper/validation/write_only.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package validation
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/go-cty/cty"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// PreferWriteOnlyAttribute is a ValidateResourceConfigFunc that returns a warning
+// if the Terraform client supports write-only attributes and the old attribute is
+// not null.
+// The last step in the path must be a cty.GetAttrStep{}.
+// When creating a cty.IndexStep{} to into a nested attribute, use an unknown value
+// of the index type to indicate any key value.
+// For lists: cty.Index(cty.UnknownVal(cty.Number)),
+// For maps: cty.Index(cty.UnknownVal(cty.String)),
+// For sets: cty.Index(cty.UnknownVal(cty.Object(nil))),
+func PreferWriteOnlyAttribute(oldAttribute cty.Path, writeOnlyAttributeName string) schema.ValidateResourceConfigFunc {
+	return func(ctx context.Context, req schema.ValidateResourceConfigFuncRequest, resp *schema.ValidateResourceConfigFuncResponse) {
+		if !req.WriteOnlyAttributesAllowed {
+			return
+		}
+
+		var oldAttrs []attribute
+
+		err := cty.Walk(req.RawConfig, func(path cty.Path, value cty.Value) (bool, error) {
+			if PathEquals(path, oldAttribute) {
+				oldAttrs = append(oldAttrs, attribute{
+					value: value,
+					path:  path,
+				})
+			}
+
+			return true, nil
+		})
+		if err != nil {
+			return
+		}
+
+		for _, attr := range oldAttrs {
+			attrPath := attr.path.Copy()
+
+			pathLen := len(attrPath)
+
+			if pathLen == 0 {
+				return
+			}
+
+			lastStep := attrPath[pathLen-1]
+
+			// Only attribute steps have a Name field
+			attrStep, ok := lastStep.(cty.GetAttrStep)
+			if !ok {
+				resp.Diagnostics = diag.Diagnostics{
+					{
+						Severity:      diag.Error,
+						Summary:       "Invalid oldAttributePath",
+						Detail:        "The specified oldAttribute path must point to an attribute.",
```

Comment:
Since this is technically an error for the provider developer and not the practitioner, we need to add to this message indicating that it should be reported to the provider developers.